Full decrypted ISO support

Functionable opened the pull request:
This PR implements complete support for playing games from decrypted ISO images in both for the emulator and the GUI. As requested in #4021 

I've tried to limit the changes to existing code where possible. 
I tried to keep the changes as contained as I could especially in the GUI bits.

The code has been only tested on Linux and Windows (x64 only for both) and it appears that games boot and play fine there.

Megamouse commented on rpcs3/rpcs3qt/savestate_manager_dialog.cpp:
```diff
@@ -445,6 +445,18 @@ void savestate_manager_dialog::ResizeGameIcons()
+							break;
+						}
+					}
+				}
```
This should not happen here. We do not want to access m_game_info in a random thread.
You'll have to figure out how to do this in StartSavestateLoadThreads

Functionable replied:
I ended up adding a new string field to game_savestates_data to accomplish this, and making the thread load the icon from the ISO archive itself. Hopefully this is better.

Megamouse commented on rpcs3/main_application.cpp:
```diff
@@ -343,7 +343,8 @@ EmuCallbacks main_application::CreateCallbacks()
 		// May result in an empty string if path does not exist
-		return QFileInfo(QString::fromUtf8(sv.data(), static_cast<int>(sv.size()))).canonicalFilePath().toStdString();
+		std::string result = QFileInfo(QString::fromUtf8(sv.data(), static_cast<int>(sv.size()))).canonicalFilePath().toStdString();
+		return !result.empty() ? result : std::string(sv);
```
this looks like a hack

Functionable replied:
Yeah it is. It seemed to be needed since a few checks in BootGame would blow up otherwise and cause games not to boot, along with Emulator::GetFakeCat returning "HG" for ISO games despite m_cat being "DG". I did some some more checks now, Emulator::BootGame works fine and I just needed to modify GetFakeCat to fix this behaviour.

elad335 commented on rpcs3/Loader/ISO.cpp:
```diff
@@ -0,0 +1,600 @@
+
+iso_fs_node* iso_archive::retrieve(const std::string& passed_path)
+{
+	std::string path = std::filesystem::path(passed_path).lexically_normal().string();
```
Well techncially, this is a bug. Real filesystem when it sees something like "/directory_exist/does_not_exist_directory/../" it fails.
You need to support "/.." properly.

elad335 commented on rpcs3/Loader/ISO.cpp:
```diff
@@ -0,0 +1,600 @@
+
+void load_iso(const std::string& path)
+{
+	fs::set_virtual_device("iso_overlay_fs_dev",
```
Seems like this line should not be here.

Functionable commented on rpcs3/rpcs3qt/game_list_frame.cpp:
```diff
@@ -708,27 +742,43 @@ void game_list_frame::OnParsingFinished()
+					game.icon = QPixmap::fromImage(iconImage);
+				}
+				game.info.icon_path.clear();
+			}
```
I've moved it there. Please tell me if this is any better now.

Functionable commented on rpcs3/Loader/ISO.cpp:
```diff
@@ -0,0 +1,600 @@
+	u64 local_remaining = local_extent_remaining(offset);
+
+	u64 total_read = m_file.read_at(file_offset(offset), buffer, std::min(size, local_remaining));
+	if (total_read == bad_res) return -1;
```
I've removed the checks and it looks a lot better now.

Functionable commented on rpcs3/Emu/System.cpp:
```diff
@@ -1406,6 +1408,10 @@ game_boot_result Emulator::Load(const std::string& title_id, bool is_disc_patch,
 			}
+			else if (is_file_iso(argv[0]))
+			{
+				m_path = argv[0];
```
It's needed to launch from save states that were done on games running from ISO. Is there some place else I should move this?

Megamouse commented on rpcs3/rpcs3qt/savestate_manager_dialog.cpp:
```diff
@@ -445,6 +448,20 @@ void savestate_manager_dialog::ResizeGameIcons()
+					{
+						icon = QPixmap::fromImage(iconImage);
+					}
+				}
```
maybe just implement in qt_utils:

bool load_icon(QPixmap& icon, const std::string& icon_path, const std::string& archive_path);

and use that in line 438 instead of icon.load(..)

You can then reuse the function in game_list_base::IconLoadFunction line 53

elad335 commented:
> <img alt="games – Datei-Explorer 12 12 2025 17_45_08" width="1159" height="1185" src="https://private-user-images.githubusercontent.com/29849893/525981853-d19e64a5-7827-4ebf-8d50-3893846dc7ee.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc0MjY3ODAsIm5iZiI6MTc2NzQyNjQ4MCwicGF0aCI6Ii8yOTg0OTg5My81MjU5ODE4NTMtZDE5ZTY0YTUtNzgyNy00ZWJmLThkNTAtMzg5Mzg0NmRjN2VlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTAzVDA3NDgwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcxNDk4ZTljZjU1ZWZjNDQ3NDAyNjVlOWEyMTFhM2VhMjFiZjczZTg1OTIzNDM2YmI2MGM1NjAyNjMxNWY1NWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ymFy1iginTJbZPIn2L2cS_iZ5C4-SUdaTjdAU_cKI8o"> <img alt="RPCS3 0 0 38-1c6b7610 Alpha _ 17855_merge 12 12 2025 17_45_23" width="1349" height="1360" src="https://private-user-images.githubusercontent.com/29849893/525981858-ec86c30e-adb9-4710-a7ed-0990f16ef1bc.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc0MjY3ODAsIm5iZiI6MTc2NzQyNjQ4MCwicGF0aCI6Ii8yOTg0OTg5My81MjU5ODE4NTgtZWM4NmMzMGUtYWRiOS00NzEwLWE3ZWQtMDk5MGYxNmVmMWJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTAzVDA3NDgwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBhNzc1YzUyZWUxMTZlNzMwMWEzOGRhYTkyYTNkMzJjZTA4MzVlYzg2OWUxN2IxNGZjZWRmZGJlMzI4NjYwMTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Av6B6F2s08tVa8_TdKcFALp1lh2nPPWZyhk6GEguDCY">
> I don't see any of the games in the rpcs3 GUI when the games are in the games folder.
> 
> But the game starts when I drag and drop them.
> 
> edit: same on Mac <img alt="Bildschirmfoto 2025-12-12 um 18 04 55" width="2000" height="1056" src="https://private-user-images.githubusercontent.com/29849893/525989040-36004805-6f48-41ca-ae22-6c6ec671661f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc0MjY3ODAsIm5iZiI6MTc2NzQyNjQ4MCwicGF0aCI6Ii8yOTg0OTg5My81MjU5ODkwNDAtMzYwMDQ4MDUtNmY0OC00MWNhLWFlMjItNmM2ZWM2NzE2NjFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTAzVDA3NDgwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYzYjMxNzZmMzBiOGIyZTAzZDAyYWUwMzVhZWQ4ZTA1MTc4ZmEwZWQ2NDU5OGM5YzQzMDJhYmE1YmE0MjE4Y2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mQGfwZGM4X_VA6dptfycYKlSqvVAenMtVRR2w8kEEZ8">

HI, what's the progress on this? did you reproduce this bug?

Florin9doi commented:
I tested this PR and I noticed few differences in comparison with extracted games: 
- Drag and drop a single game to the window:
  - Extracted game: the game is added to library
  - ISO: the game is added to library and started automatically
- Drag and drop multiple games:
  - Extracted games: all are added to library
  - ISO: not possible
- File -> Add games -> select a folder containing both extracted and ISO games. Only the extracted games are added to library.

Functionable commented:
I rebased due to some problems I had in the workflows building for mac. I have also fixed a few more bugs I found when testing things out. @darkuser89 @Florin9doi, I have fixed the problems you've identified in your comments. Could you please test these changes if you have the time?

Megamouse commented on rpcs3/rpcs3qt/qt_utils.cpp:
```diff
@@ -682,6 +683,29 @@ namespace gui
+			icon_file.read(data.data(), icon_size);
+			QImage iconImage;
+
+			if (iconImage.loadFromData(data))
```
why not just:

```suggestion
			if (!icon.loadFromData(data))
			{
				icon = {};
				return false;
			}
```

Megamouse commented on rpcs3/rpcs3qt/qt_utils.cpp:
```diff
@@ -682,6 +683,29 @@ namespace gui
 		}
 
+		bool load_icon(QPixmap& icon, const std::string& icon_path,
+			const std::string& archive_path)
```
```suggestion
		bool load_icon(QPixmap& icon, const std::string& icon_path, const std::string& archive_path)
```

Florin9doi commented:
I confirm that all the problems I mentioned above are now resolved
